### PR TITLE
Fix File updating

### LINKED ISSUE #3335: Cached Examples don't work for File component
### Describe the bug

Using cached examples for `gr.File` input component or output components does not work.

1. The input files cannot be downloaded
2. Clicking `Submit` with the cached input results in an error

The cached output file can be downloaded and contains the expected output.

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Reproduction

Run the `zip_files` demo.

I tried with the duplicated `titanic.csv` files as well as different files for each input and it made no difference.

### Screenshot

cached_examples_file_component

### Logs

```shell
Traceback (most recent call last):
  File "/Users/freddy/sources/gradio/gradio/routes.py", line 384, in run_predict
    output = await app.get_blocks().process_api(
  File "/Users/freddy/sources/gradio/gradio/blocks.py", line 1022, in process_api
    inputs = self.preprocess_data(fn_index, inputs, state)
  File "/Users/freddy/sources/gradio/gradio/blocks.py", line 914, in preprocess_data
    processed_input.append(block.preprocess(inputs[i]))
  File "/Users/freddy/sources/gradio/gradio/components.py", line 2415, in preprocess
    return [process_single_file(f) for f in x]
  File "/Users/freddy/sources/gradio/gradio/components.py", line 2415, in <listcomp>
    return [process_single_file(f) for f in x]
  File "/Users/freddy/sources/gradio/gradio/components.py", line 2377, in process_single_file
    f["name"],
KeyError: 'name'
```

### System Info

```shell
main
```

### Severity

serious, but I can work around it

### LINKED ISSUE #3365: error when downloading files: RuntimeError: File at path /content/undefined does not exist.
### Describe the bug

When I upload a file and click submit, the uploaded file appears with a path like "/tmp/filename". When I then try to click on "download", a new browser tab opens and an error appears: `Internal Server Error`.

I experience this when uploading + processing a file and then wanting to download it again from the same file component. 

The weird thing is that the same code worked a month ago with an older version of gradio and I've also tried it with the latest 3.19. I have the impression that this always happens when the displayed file path is `/temp/filename` instead of just `/filename`

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Reproduction

The same issue also appears with the example code for File in the docs, see here:  https://gradio.app/docs/#file-header
```
import os
from zipfile import ZipFile

import gradio as gr


def zip_files(files):
    with ZipFile("tmp.zip", "w") as zipObj:
        for idx, file in enumerate(files):
            zipObj.write(file.name, file.name.split("/")[-1])
    return "tmp.zip"

demo = gr.Interface(
    zip_files,
    gr.File(file_count="multiple", file_types=["text", ".json", ".csv"]),
    "file",
    examples=[[[os.path.join(os.path.dirname(__file__),"files/titanic.csv"), 
    os.path.join(os.path.dirname(__file__),"files/titanic.csv"), 
    os.path.join(os.path.dirname(__file__),"files/titanic.csv")]]], 
    cache_examples=True
)

if __name__ == "__main__":
    demo.launch()
```

### Screenshot

screenshot

new tab that pops up with the error when clicking on download (left column) in the app from the official docs: 
Screenshot 2023-03-02 at 16 01 26

### Logs

Stack trace from my private/local app: 

```shell
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/starlette/responses.py", line 335, in __call__
    stat_result = await anyio.to_thread.run_sync(os.stat, self.path)
  File "/usr/local/lib/python3.8/dist-packages/anyio/to_thread.py", line 31, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
  File "/usr/local/lib/python3.8/dist-packages/anyio/_backends/_asyncio.py", line 937, in run_sync_in_worker_thread
    return await future
  File "/usr/local/lib/python3.8/dist-packages/anyio/_backends/_asyncio.py", line 867, in run
    result = context.run(func, *args)
FileNotFoundError: [Errno 2] No such file or directory: '/content/undefined'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/uvicorn/protocols/http/h11_impl.py", line 407, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "/usr/local/lib/python3.8/dist-packages/uvicorn/middleware/proxy_headers.py", line 78, in __call__
    return await self.app(scope, receive, send)
  File "/usr/local/lib/python3.8/dist-packages/fastapi/applications.py", line 271, in __call__
    await super().__call__(scope, receive, send)
  File "/usr/local/lib/python3.8/dist-packages/starlette/applications.py", line 118, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/usr/local/lib/python3.8/dist-packages/starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "/usr/local/lib/python3.8/dist-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/usr/local/lib/python3.8/dist-packages/starlette/middleware/cors.py", line 84, in __call__
    await self.app(scope, receive, send)
  File "/usr/local/lib/python3.8/dist-packages/starlette/middleware/exceptions.py", line 79, in __call__
    raise exc
  File "/usr/local/lib/python3.8/dist-packages/starlette/middleware/exceptions.py", line 68, in __call__
    await self.app(scope, receive, sender)
  File "/usr/local/lib/python3.8/dist-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__
    raise e
  File "/usr/local/lib/python3.8/dist-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
    await self.app(scope, receive, send)
  File "/usr/local/lib/python3.8/dist-packages/starlette/routing.py", line 706, in __call__
    await route.handle(scope, receive, send)
  File "/usr/local/lib/python3.8/dist-packages/starlette/routing.py", line 276, in handle
    await self.app(scope, receive, send)
  File "/usr/local/lib/python3.8/dist-packages/starlette/routing.py", line 69, in app
    await response(scope, receive, send)
  File "/usr/local/lib/python3.8/dist-packages/starlette/responses.py", line 338, in __call__
    raise RuntimeError(f"File at path {self.path} does not exist.")
RuntimeError: File at path /content/undefined does not exist.
```

### System Info

```shell
Google colab, 
gradio 3.19
```

### Severity

serious, but I can work around it

### PR DESCRIPTION
**Pull request by freddyaboulton**
# Description

Fixes #3365 
Fixes #3335

Thanks @pngwn for helping me with the unit test 🙏 

Additional testing:

1. Run the repro from #3365 

```python
import gradio as gr
import tempfile
import pathlib

def translate(input_file):
    new_name = str(pathlib.Path(input_file.name).with_name("processed"))
    with tempfile.NamedTemporaryFile(mode="w",
            prefix=new_name,
            delete=False) as f:
        f.write(open(input_file.name).read())
    return f.name, f.name

with gr.Blocks() as demo:
    with gr.Row():
        input_file = gr.File(label="Input")
        output_file = gr.File(label="Output")
    process = gr.Button(value="Process")
    
    process.click(translate, inputs=[input_file], outputs=[input_file, output_file])


demo.launch()
```

Both files should be downloadable after processing.
file_repro

2. Run zip_files with example caching

Hitting `submit` after clicking on the examples should preserve the name and you should be able to download them. Can verify this on the deployed space.

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have added a short summary of my change to the CHANGELOG.md
- [x] My code follows the style guidelines of this project
- [x] I have commented my code in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

# A note about the CHANGELOG

Hello 👋 and thank you for contributing to Gradio!

All pull requests must update the change log located in CHANGELOG.md, unless the pull request is labeled with the "no-changelog-update" label.

Please add a brief summary of the change to the Upcoming Release > Full Changelog section of the CHANGELOG.md file and include
a link to the PR (formatted in markdown) and a link to your github profile (if you like). For example, "* Added a cool new feature by `[@myusername](link-to-your-github-profile)` in `[PR 11111](https://github.com/gradio-app/gradio/pull/11111)`".

If you would like to elaborate on your change further, feel free to include a longer explanation in the other sections.
If you would like an image/gif/video showcasing your feature, it may be best to edit the CHANGELOG file using the 
GitHub web UI since that lets you upload files directly via drag-and-drop.

### CODE REVIEW

**Comment by abidlabs** [comment]
Tested and works great @freddyaboulton! I would get a review from a frontend person too

**Review by abidlabs — APPROVED** [review]

File: ui/packages/app/src/components/File/File.test.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+	// 	const user_2 = getAllByTestId("user");
+	// 	const bot_2 = getAllByTestId("bot");
+
+	// 	assert.equal(user_2.length, 2);
+	// 	assert.equal(bot_2.length, 2);
+
+	// 	assert.equal(get_text(user_2[1]), "user message two");
+	// 	assert.equal(get_text(bot_2[1]), "bot message two");
+	// });
```
Comment (pngwn) [inline ui/packages/app/src/components/File/File.test.ts:93]: Leftover.

**Review by pngwn — APPROVED** [review]
Tested and working great. Nice test @freddyaboulton! Since you seem so keen you can now take over all testing duties.

lgtm!

**Comment by freddyaboulton** [comment]
Thank you for the reviews @pngwn and @abidlabs !

**Pull request merged**